src/cloudsc_gpu/cloudsc_driver_gpu_omp_scc_hoist_mod.F90

```diff
@@ -0,0 +1,236 @@
+    REAL(KIND=JPRB), INTENT(OUT) :: PFHPSL(NPROMA, NLEV+1, NGPBLKS)    ! Enthalpy flux for liq
+    REAL(KIND=JPRB), INTENT(OUT) :: PFHPSN(NPROMA, NLEV+1, NGPBLKS)    ! ice number concentration (cf. CCN)
+
+    ! Local declarations of promoted temporaries
+    REAL(KIND=JPRB) :: ZFOEALFA(NPROMA, NLEV+1, NGPBLKS)
+    REAL(KIND=JPRB) :: ZTP1(NPROMA, NLEV, NGPBLKS)
+    REAL(KIND=JPRB) :: ZLI(NPROMA, NLEV, NGPBLKS)
+    REAL(KIND=JPRB) :: ZA(NPROMA, NLEV, NGPBLKS)
+    REAL(KIND=JPRB) :: ZAORIG(NPROMA, NLEV, NGPBLKS)
+    REAL(KIND=JPRB) :: ZLIQFRAC(NPROMA, NLEV, NGPBLKS)
+    REAL(KIND=JPRB) :: ZICEFRAC(NPROMA, NLEV, NGPBLKS)
+    REAL(KIND=JPRB) :: ZQX(NPROMA, NLEV, NCLV, NGPBLKS)
+    REAL(KIND=JPRB) :: ZQX0(NPROMA, NLEV, NCLV, NGPBLKS)
+    REAL(KIND=JPRB) :: ZPFPLSX(NPROMA, NLEV+1, NCLV, NGPBLKS)
+    REAL(KIND=JPRB) :: ZLNEG(NPROMA, NLEV, NCLV, NGPBLKS)
+    REAL(KIND=JPRB) :: ZQXN2D(NPROMA, NLEV, NCLV, NGPBLKS)
+    REAL(KIND=JPRB) :: ZQSMIX(NPROMA, NLEV, NGPBLKS)
+    REAL(KIND=JPRB) :: ZQSLIQ(NPROMA, NLEV, NGPBLKS)
+    REAL(KIND=JPRB) :: ZQSICE(NPROMA, NLEV, NGPBLKS)
+    REAL(KIND=JPRB) :: ZFOEEWMT(NPROMA, NLEV, NGPBLKS)
+    REAL(KIND=JPRB) :: ZFOEEW(NPROMA, NLEV, NGPBLKS)
+    REAL(KIND=JPRB) :: ZFOEELIQT(NPROMA, NLEV, NGPBLKS)
+    INTEGER(KIND=JPIM) :: JL
+
+    INTEGER(KIND=JPIM) :: JKGLO,IBL,ICEND
+    TYPE(PERFORMANCE_TIMER) :: TIMER
+    INTEGER(KIND=JPIM) :: TID ! thread id from 0 .. NUMOMP - 1
+
+    ! Local copy of cloud parameters for offload
+    TYPE(TECLDP) :: LOCAL_YRECLDP
+
+    NGPBLKS = (NGPTOT / NPROMA) + MIN(MOD(NGPTOT,NPROMA), 1)
+1003 format(5x,'NUMPROC=',i0,', NUMOMP=',i0,', NGPTOTG=',i0,', NPROMA=',i0,', NGPBLKS=',i0)
+    if (irank == 0) then
+      write(0,1003) NUMPROC,NUMOMP,NGPTOTG,NPROMA,NGPBLKS
+    end if
+
+    ! Global timer for the parallel region
+    CALL TIMER%START(NUMOMP)
+
+!$omp target enter data map(alloc: ZFOEALFA, ZTP1, ZLI, ZA, ZAORIG, ZLIQFRAC, ZICEFRAC, ZQX, ZQX0,  &
+!$omp &   ZPFPLSX, ZLNEG, ZQXN2D, ZQSMIX, ZQSLIQ, ZQSICE, ZFOEEWMT,  &
+!$omp &   ZFOEEW, ZFOEELIQT)
+
+
+    ! Workaround for PGI / OpenACC oddities:
+    ! Create a local copy of the parameter struct to ensure they get
+    ! moved to the device the in ``acc data`` clause below
+    LOCAL_YRECLDP = YRECLDP
+
+!$omp target data &
+!$omp map(to: &
+!$omp   pt,pq,buffer_cml,buffer_tmp,pvfa, &
+!$omp   pvfl,pvfi,pdyna,pdynl,pdyni,phrsw,phrlw,pvervel, &
+!$omp   pap,paph,plsm,ldcum,ktype,plu,psnde, &
+!$omp   pmfu,pmfd,pa,pclv,psupsat,plcrit_aer,picrit_aer, &
+!$omp   pre_ice,pccn,pnice, yrecldp) &
+!$omp map(tofrom: &
+!$omp   buffer_loc,plude,pcovptot,prainfrac_toprfz) &
+!$omp map(from: &
+!$omp   pfsqlf,pfsqif,pfcqnng, &
+!$omp   pfcqlng ,pfsqrf,pfsqsf,pfcqrng,pfcqsng,pfsqltur, &
+!$omp   pfsqitur,pfplsl,pfplsn,pfhpsl,pfhpsn)
+
+    ! Local timer for each thread
+    TID = GET_THREAD_NUM()
+    CALL TIMER%THREAD_START(TID)
+
+#ifdef HAVE_OMP_TARGET_LOOP_CONSTRUCT
```

Comment:
It is standard version compliance-specific behaviour. The loop construct was added with OpenMP 5.0 but not all compilers implement it. We switch between the versions automatically by compiling snippets with loop construct and target teams distribute in the CMake configuration. If the first is supported, we use that, otherwise fall back to the second, and switch off this variant completely if neither works.

Reply:
I'll mention it in the README, though